### ISSUE TITLE
Extract Project ID from Service Account JSON File

### ISSUE THREAD

**Issue by yasmanillanes** [calalalizade/fcm-rs#1]
This is more of a suggestion, really.

Have you considered extracting the Project ID from the Service Account JSON file instead of requiring it as a parameter in the `FCMClient`? I know it's nit-picky, but the developer wouldn't have to go digging for yet another piece of information off the Firebase Console.

Extracting the Project ID from the JSON file would be quite trivial. Its key is labeled `project_id` right at the root of the file.

**Comment by calalalizade** [comment]
Thank you so much for the suggestion! I appreciate the idea of extracting the Project ID directly from the JSON file. It makes perfect sense and would definitely simplify the setup process.

I'll look into implementing this feature as soon as I have some time. In the meantime, if you’d like to try adding this yourself, please feel free to send a pull request.

**Comment by yasmanillanes** [comment]
Hi Jalal,

I actually already got started on this. I branched out off your repository and implemented this feature, plus the ability for the library consumer to see error responses from FCM, which they couldn't before. Seeing FCM error responses is kind of important, especially when the error is due to an invalidated token; the API consumer needs to know which tokens they need to remove from their database.

I just need to test my code first and will be creating a pull request soon.

Thanks.

-------- Message d'origine --------

**Comment by calalalizade** [comment]
Thanks for starting on this. Take your time to test the code, and then we can move forward with the pull request. Again, appreciate your effort.

**Comment by yasmanillanes** [comment]
Hi @calalalizade

I created my pull request. Some caveats, though:

I was only able to test the "FCM error response" path I added to the code. I'm in the middle of building my server that uses your crate and don't have the mobile client app yet to generate actual valid device tokens to test the "FCM success response" path. I assume you have a mobile client app somewhere you can use to quickly test this. Otherwise, I still have to develop my mobile client app and that's still quite a ways to go. I was hoping you could get a head start on that part of the testing. I'm pretty sure I didn't break anything since I changed next to nothing on the "success" path, but you never know.

I made some breaking changes to the code; that is, consumers of the crate who migrate from version 0.1.0 to version 0.2.0 will have to modify their code slightly (mainly, stop passing the `project_id` parameter to the `Client` and reference the new `FcmSuccessResponse` instead of the now removed `FcmSendResponse` struct. If you'd like to keep the crate backwards compatible, it's only a matter of adding a type alias and overload a couple of functions in the code.

I'm a bit inexperienced with Rust still. If you know of a smarter way to implement any of the changes I submitted, let me know.

Thanks.

**Comment by calalalizade** [comment]
Hi @yasmanillanes,

Thank you for your pull request and for providing detailed information about the changes. I’ve reviewed your pull request and the modifications you made. Everything seems to be in order for now. I understand the limitations regarding the testing of the success response path. Unfortunately, I don’t have a mobile client app ready for testing at the moment, but I’m also in the process of developing one. I’ll ensure that this part is tested as soon as the mobile client is up and running.

If you have any further suggestions or if there are specific areas where you think improvements could be made, please let me know. Thanks again for your contribution, and I’ll keep you updated on the progress once we’ve had the chance to test the success response.

**Comment by calalalizade** [comment]
Hi @yasmanillanes,

Thank you for your pull request and for providing detailed information about the changes. I’ve reviewed your pull request and the modifications you made. Everything seems to be in order for now. I understand the limitations regarding the testing of the success response path. Unfortunately, I don’t have a mobile client app ready for testing at the moment, but I’m also in the process of developing one. I’ll ensure that this part is tested as soon as the mobile client is up and running.

If you have any further suggestions or if there are specific areas where you think improvements could be made, please let me know. Thanks again for your contribution, and I’ll keep you updated on the progress once we’ve had the chance to test the success response.

**Comment by calalalizade** [comment]
Hi @yasmanillanes,

Thank you for your pull request and for providing detailed information about the changes. I’ve reviewed your pull request and the modifications you made. Everything seems to be in order for now. I understand the limitations regarding the testing of the success response path. Unfortunately, I don’t have a mobile client app ready for testing at the moment, but I’m also in the process of developing one. I’ll ensure that this part is tested as soon as the mobile client is up and running.

If you have any further suggestions or if there are specific areas where you think improvements could be made, please let me know. Thanks again for your contribution, and I’ll keep you updated on the progress once we’ve had the chance to test the success response.

**Comment by yasmanillanes** [comment]
Hi @calalalizade ,

I finally got to the point on the client side where I could test the success response path and it all seems to be working as intended. The server gets the proper response from FCM and the client gets a proper message sent by the server via FCM.

Right now I'm just referencing the source code of the fork I created instead of your crate on crates.io, so whenever you're ready to compile these changes onto your official crate, let me know so that I can make the switch 👍 

Thank you very much!
